src/dfx/src/commands/canister/metadata.rs

```diff
@@ -0,0 +1,41 @@
+use crate::lib::error::DfxResult;
+use crate::lib::models::canister_id_store::CanisterIdStore;
+use crate::lib::root_key::fetch_root_key_if_needed;
+use crate::Environment;
+
+use anyhow::{anyhow, Context};
+use clap::Parser;
+use ic_types::Principal;
+use std::io::{stdout, Write};
+
+/// Sign a canister call and generate message file.
+#[derive(Parser)]
+pub struct CanisterMetadataOpts {
+    /// Specifies the name of the canister to call.
+    canister_name: String,
+
+    /// Specifies the name of the metadata to retrieve.
+    metadata_name: String,
+}
+
+pub async fn exec(env: &dyn Environment, opts: CanisterMetadataOpts) -> DfxResult {
+    let agent = env
+        .get_agent()
+        .ok_or_else(|| anyhow!("Cannot get HTTP client from environment."))?;
+
+    let callee_canister = opts.canister_name.as_str();
+    let canister_id_store = CanisterIdStore::for_env(env)?;
+
+    let canister_id = Principal::from_text(callee_canister)
+        .or_else(|_| canister_id_store.get(callee_canister))?;
+
+    fetch_root_key_if_needed(env).await?;
+    let metadata = agent
+        .read_state_canister_metadata(canister_id, &opts.metadata_name, false)
+        .await
+        .with_context(|| format!("Failed to read controllers of canister {}.", canister_id))?;
+
+    stdout().write_all(&metadata)?;
```

Comment:
Not sure the best way to display this. Some of the metadata may not be printable ascii. Maybe we can check if `metadata` is printable, if not, we display the hex instead.

Reply:
Would it make sense to have an `--outfile` flag instead/in addition?

Reply:
@chenyan-dfinity @sesi200 I wondered about these too.  Not sure of the best way either.

The thinking was that by writing the raw data bytes to stdout, they can be written to a file or piped to another program.

While the metadata may not be printable ascii, what if the developer wants to run the actual contents through `hexdump` or some other tool?  If `dfx canister metadata ... | hexdump -C`, and the metadata is not entirely printable ascii, what would the expected behavior be?

What would the proposed rules be?  Would they more clear than always writing as-is to stdout?  The below seem overly complicated to me.
- if `--outfile` specified, write data to specified file as-is.  `--outfile -` means write data as-is to stdout.
- If no `--outfile` specified, and metadata is printable ascii, write that to stdout.
- If no `--outfile` specified, and metadata is not all printable ascii, then write an explanation to stderr and display as hex.

Reply:
I have no strong opinion either way and have no problem merging this as-is. I was mostly thinking about invoking dfx from somewhere that doesn't have such trivial redirect mechanics, but I might just be overthinking it.

My approach would have been to print to `--outfile` if specified, otherwise print to stdout. Reason being that I'd want to give a way to not see garbage printed if it is expected.